From 3ecc4a1a1948ba7a3ef8ec1dfb7951f1d7b2bc0b Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 18 Jul 2026 00:04:47 -0400
Subject: [PATCH 198/268] winewayland/win32u: Improve colorspace conversion.

This does everything except expose the new colorspaces in vkGetPhysicalDeviceSurfaceCapabilities(2)KHR
---
 dlls/win32u/vulkan.c                   | 59 ++++++++++++++++----
 dlls/winemac.drv/vulkan.c              |  6 +--
 dlls/winewayland.drv/vulkan.c          | 39 +++++++-------
 dlls/winewayland.drv/wayland.c         | 59 +++++++++++++++++---
 dlls/winewayland.drv/wayland_output.c  | 74 +-------------------------
 dlls/winewayland.drv/wayland_surface.c |  1 +
 dlls/winewayland.drv/waylanddrv.h      |  4 +-
 dlls/winex11.drv/vulkan.c              |  6 +--
 include/wine/vulkan_driver.h           |  2 +-
 9 files changed, 133 insertions(+), 117 deletions(-)

diff --git a/dlls/win32u/vulkan.c b/dlls/win32u/vulkan.c
index c4153e2ce46..eeea5b517d9 100644
--- a/dlls/win32u/vulkan.c
+++ b/dlls/win32u/vulkan.c
@@ -1965,7 +1965,7 @@ static VkResult win32u_vkGetPhysicalDeviceSurfaceFormatsKHR( VkPhysicalDevice cl
     struct vulkan_instance *instance = physical_device->instance;
 
     return instance->p_vkGetPhysicalDeviceSurfaceFormatsKHR( physical_device->host.physical_device,
-                                                                surface->obj.host.surface, format_count, formats );
+                                                             surface->obj.host.surface, format_count, formats );
 }
 
 static VkResult win32u_vkGetPhysicalDeviceSurfaceFormats2KHR( VkPhysicalDevice client_physical_device, const VkPhysicalDeviceSurfaceInfo2KHR *surface_info,
@@ -2540,6 +2540,8 @@ static VkResult win32u_vkCreateSwapchainKHR( VkDevice client_device, const VkSwa
     VkSwapchainCreateInfoKHR create_info_host = *create_info;
     VkSurfaceCapabilitiesKHR capabilities;
     VkSwapchainKHR host_swapchain;
+    uint32_t format_count = 0;
+    VkSurfaceFormatKHR *formats;
     RECT client_rect;
     VkResult res;
 
@@ -2556,6 +2558,7 @@ static VkResult win32u_vkCreateSwapchainKHR( VkDevice client_device, const VkSwa
     res = instance->p_vkGetPhysicalDeviceSurfaceCapabilitiesKHR( physical_device->host.physical_device, surface->obj.host.surface, &capabilities );
     if (res) return res;
 
+    create_info_host.imageColorSpace = driver_funcs->p_vulkan_map_colorspace( create_info_host.imageColorSpace, surface->client );
     create_info_host.imageExtent.width = max( create_info_host.imageExtent.width, capabilities.minImageExtent.width );
     create_info_host.imageExtent.height = max( create_info_host.imageExtent.height, capabilities.minImageExtent.height );
 
@@ -2589,10 +2592,46 @@ static VkResult win32u_vkCreateSwapchainKHR( VkDevice client_device, const VkSwa
                    create_info_host.imageFormat );
     }
 
-    if ((res = driver_funcs->p_vulkan_colorspace_configure( &create_info_host.imageColorSpace, surface->client )))
+    /* check if the new colorspace works with the provided format */
+    if (create_info_host.imageColorSpace != create_info->imageColorSpace)
     {
-        free( swapchain );
-        return res;
+        BOOL found = FALSE;
+
+        res = instance->p_vkGetPhysicalDeviceSurfaceFormatsKHR( physical_device->host.physical_device, surface->obj.host.surface, &format_count, NULL );
+
+        if (!(formats = calloc( format_count, sizeof(*formats) ))) return VK_ERROR_OUT_OF_HOST_MEMORY;
+
+        res = instance->p_vkGetPhysicalDeviceSurfaceFormatsKHR( physical_device->host.physical_device, surface->obj.host.surface, &format_count, formats );
+
+        if (res)
+        {
+            free(formats);
+            return res;
+        }
+
+    again:
+        for (unsigned i = 0; i < format_count; i++)
+        {
+            if (formats[i].format == create_info_host.imageFormat &&
+                formats[i].colorSpace == create_info_host.imageColorSpace)
+                found = TRUE;
+        }
+
+        /* HACK: try again with VK_COLOR_SPACE_SRGB_NONLINEAR_KHR */
+        if (!found && create_info_host.imageColorSpace == VK_COLOR_SPACE_PASS_THROUGH_EXT)
+        {
+            create_info_host.imageColorSpace = VK_COLOR_SPACE_SRGB_NONLINEAR_KHR;
+            goto again;
+        }
+
+        if (!found)
+        {
+            ERR("Colorspace %u is not compatible with format %u\n",
+                create_info_host.imageColorSpace, create_info_host.imageFormat);
+            create_info_host.imageColorSpace = create_info->imageColorSpace;
+        }
+
+        free(formats);
     }
 
     InterlockedIncrement( &surface->client->busy_ref );
@@ -4013,9 +4052,9 @@ static VkResult nulldrv_vulkan_surface_create( HWND hwnd, BOOL raw, const struct
     return res;
 }
 
-static VkResult nulldrv_vulkan_colorspace_configure( VkColorSpaceKHR *colorspace, struct client_surface *client )
+static VkColorSpaceKHR nulldrv_vulkan_map_colorspace( VkColorSpaceKHR colorspace, struct client_surface *client )
 {
-    return VK_SUCCESS;
+    return colorspace;
 }
 
 static VkBool32 nulldrv_get_physical_device_presentation_support( struct vulkan_physical_device *physical_device, uint32_t queue )
@@ -4044,7 +4083,7 @@ static void nulldrv_map_device_extensions( struct vulkan_device_extensions *exte
 static const struct vulkan_driver_funcs nulldrv_funcs =
 {
     .p_vulkan_surface_create = nulldrv_vulkan_surface_create,
-    .p_vulkan_colorspace_configure = nulldrv_vulkan_colorspace_configure,
+    .p_vulkan_map_colorspace = nulldrv_vulkan_map_colorspace,
     .p_get_physical_device_presentation_support = nulldrv_get_physical_device_presentation_support,
     .p_map_instance_extensions = nulldrv_map_instance_extensions,
     .p_map_device_extensions = nulldrv_map_device_extensions,
@@ -4077,10 +4116,10 @@ static VkResult lazydrv_vulkan_surface_create( HWND hwnd, BOOL raw, const struct
     return driver_funcs->p_vulkan_surface_create( hwnd, raw, instance, surface, client );
 }
 
-static VkResult lazydrv_vulkan_colorspace_configure( VkColorSpaceKHR *colorspace, struct client_surface *client )
+static VkColorSpaceKHR lazydrv_vulkan_map_colorspace( VkColorSpaceKHR colorspace, struct client_surface *client )
 {
     vulkan_driver_load();
-    return driver_funcs->p_vulkan_colorspace_configure( colorspace, client );
+    return driver_funcs->p_vulkan_map_colorspace( colorspace, client );
 }
 
 static VkBool32 lazydrv_get_physical_device_presentation_support( struct vulkan_physical_device *physical_device, uint32_t queue )
@@ -4104,7 +4143,7 @@ static void lazydrv_map_device_extensions( struct vulkan_device_extensions *exte
 static const struct vulkan_driver_funcs lazydrv_funcs =
 {
     .p_vulkan_surface_create = lazydrv_vulkan_surface_create,
-    .p_vulkan_colorspace_configure = lazydrv_vulkan_colorspace_configure,
+    .p_vulkan_map_colorspace = lazydrv_vulkan_map_colorspace,
     .p_get_physical_device_presentation_support = lazydrv_get_physical_device_presentation_support,
     .p_map_instance_extensions = lazydrv_map_instance_extensions,
     .p_map_device_extensions = lazydrv_map_device_extensions,
diff --git a/dlls/winemac.drv/vulkan.c b/dlls/winemac.drv/vulkan.c
index d9c83675e3b..187824d10d0 100644
--- a/dlls/winemac.drv/vulkan.c
+++ b/dlls/winemac.drv/vulkan.c
@@ -90,9 +90,9 @@ err:
     return VK_ERROR_INCOMPATIBLE_DRIVER;
 }
 
-static VkResult macdrv_vulkan_colorspace_configure(VkColorSpaceKHR *colorspace, struct client_surface *client)
+static VkColorSpaceKHR macdrv_vulkan_map_colorspace(VkColorSpaceKHR colorspace, struct client_surface *client)
 {
-    return VK_SUCCESS;
+    return colorspace;
 }
 
 static VkBool32 macdrv_get_physical_device_presentation_support(struct vulkan_physical_device *physical_device,
@@ -126,7 +126,7 @@ static void macdrv_map_device_extensions(struct vulkan_device_extensions *extens
 static const struct vulkan_driver_funcs macdrv_vulkan_driver_funcs =
 {
     .p_vulkan_surface_create = macdrv_vulkan_surface_create,
-    .p_vulkan_colorspace_configure = macdrv_vulkan_colorspace_configure,
+    .p_vulkan_map_colorspace = macdrv_vulkan_map_colorspace,
     .p_get_physical_device_presentation_support = macdrv_get_physical_device_presentation_support,
     .p_map_instance_extensions = macdrv_map_instance_extensions,
     .p_map_device_extensions = macdrv_map_device_extensions,
diff --git a/dlls/winewayland.drv/vulkan.c b/dlls/winewayland.drv/vulkan.c
index 54e22f142dd..c2439a78143 100644
--- a/dlls/winewayland.drv/vulkan.c
+++ b/dlls/winewayland.drv/vulkan.c
@@ -140,44 +140,47 @@ static void wayland_map_device_extensions(struct vulkan_device_extensions *exten
     if (extensions->has_VK_KHR_external_fence_fd) extensions->has_VK_KHR_external_fence_win32 = 1;
 }
 
-static VkResult wayland_vulkan_colorspace_configure(VkColorSpaceKHR *colorspace, struct client_surface *client)
+static VkColorSpaceKHR wayland_vulkan_map_colorspace(VkColorSpaceKHR colorspace, struct client_surface *client)
 {
     struct wp_image_description_v1 *wp_image_description_v1 = NULL;
-    VkColorSpaceKHR old = *colorspace;
+    VkColorSpaceKHR new = colorspace;
 
-    if (process_wayland.supports_scrgb &&
-        old == VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT)
+    if (process_wayland.supports_win_scrgb && new == VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT)
+        new = VK_COLOR_SPACE_PASS_THROUGH_EXT;
+    else if (process_wayland.supports_win_pq && colorspace == VK_COLOR_SPACE_HDR10_ST2084_EXT)
+        new = VK_COLOR_SPACE_PASS_THROUGH_EXT;
+
+    if (!client) return new;
+    if (new == colorspace) return colorspace;
+
+    switch(colorspace)
     {
-        *colorspace = VK_COLOR_SPACE_PASS_THROUGH_EXT;
+    case VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT:
         wp_image_description_v1 =
             wp_color_manager_v1_create_windows_scrgb(process_wayland.wp_color_manager_v1);
-
-        if (!wp_image_description_v1) goto err;
-    }
-    else if (process_wayland.supports_win_pq &&
-             old == VK_COLOR_SPACE_HDR10_ST2084_EXT)
-    {
-        *colorspace = VK_COLOR_SPACE_PASS_THROUGH_EXT;
+        break;
+    case VK_COLOR_SPACE_HDR10_ST2084_EXT:
         wp_image_description_v1 =
             wp_color_manager_v1_create_windows_bt2100(process_wayland.wp_color_manager_v1);
-
-        if (!wp_image_description_v1) goto err;
+    default: break;
     }
 
-    TRACE("mapping colorspace %u => %u\n", old, *colorspace);
+    if (!wp_image_description_v1) goto err;
+
+    TRACE("mapping colorspace %u => %u\n", colorspace, new);
 
     wayland_client_surface_attach_image_description(client, wp_image_description_v1);
 
-    return VK_SUCCESS;
+    return new;
 err:
     ERR("Failed to configure image description for client surface!\n");
-    return VK_ERROR_OUT_OF_HOST_MEMORY;
+    return colorspace;
 }
 
 static const struct vulkan_driver_funcs wayland_vulkan_driver_funcs =
 {
     .p_vulkan_surface_create = wayland_vulkan_surface_create,
-    .p_vulkan_colorspace_configure = wayland_vulkan_colorspace_configure,
+    .p_vulkan_map_colorspace = wayland_vulkan_map_colorspace,
     .p_get_physical_device_presentation_support = wayland_get_physical_device_presentation_support,
     .p_map_instance_extensions = wayland_map_instance_extensions,
     .p_map_device_extensions = wayland_map_device_extensions,
diff --git a/dlls/winewayland.drv/wayland.c b/dlls/winewayland.drv/wayland.c
index 05e8f26a7c9..c1fde1db6e0 100644
--- a/dlls/winewayland.drv/wayland.c
+++ b/dlls/winewayland.drv/wayland.c
@@ -43,10 +43,6 @@ struct wayland process_wayland =
     .data_device.mutex = PTHREAD_MUTEX_INITIALIZER,
     .output_list = {&process_wayland.output_list, &process_wayland.output_list},
     .output_mutex = PTHREAD_MUTEX_INITIALIZER,
-    .supports_extended_volume = FALSE,
-    .supports_pq = FALSE,
-    .supports_scrgb = FALSE,
-    .supports_win_pq = FALSE
 };
 
 /**********************************************************************
@@ -97,6 +93,49 @@ static const struct wl_seat_listener seat_listener =
     wl_seat_handle_name
 };
 
+static void wayland_color_manager_handle_supported_intent(void *data,
+    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t intent)
+{
+}
+
+static void wayland_color_manager_handle_supported_feature(void *data,
+    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t feature)
+{
+    pthread_mutex_lock(&process_wayland.output_mutex);
+
+    TRACE("feature %u\n", feature);
+
+    if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_SCRGB)
+        process_wayland.supports_win_scrgb = TRUE;
+    else if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_BT2100)
+        process_wayland.supports_win_pq = TRUE;
+
+    pthread_mutex_unlock(&process_wayland.output_mutex);
+}
+
+static void wayland_color_manager_handle_supported_named_tf(void *data,
+    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t tf)
+{
+}
+
+static void wayland_color_manager_handle_supported_primaries(void *data,
+    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t primaries)
+{
+}
+
+static void wayland_color_manager_handle_done(void *data,
+                        struct wp_color_manager_v1 *wp_color_manager_v1)
+{
+}
+
+static const struct wp_color_manager_v1_listener wp_color_manager_listener = {
+    wayland_color_manager_handle_supported_intent,
+    wayland_color_manager_handle_supported_feature,
+    wayland_color_manager_handle_supported_named_tf,
+    wayland_color_manager_handle_supported_primaries,
+    wayland_color_manager_handle_done
+};
+
 static int wayland_disable_ssd(void)
 {
     static int disabled = -1;
@@ -270,8 +309,8 @@ static void registry_handle_global(void *data, struct wl_registry *registry,
         process_wayland.wp_color_manager_v1 =
             wl_registry_bind(registry, id, &wp_color_manager_v1_interface,
                              version < 3 ? version : 3);
-
-        wayland_color_manager_init();
+        wp_color_manager_v1_add_listener(process_wayland.wp_color_manager_v1,
+                                         &wp_color_manager_listener, NULL);
         /* Add image descriptions to existing outputs. */
         wl_list_for_each(output, &process_wayland.output_list, link)
             wayland_output_use_image_description(output);
@@ -443,7 +482,13 @@ BOOL wayland_process_init(void)
         WARN("Wayland compositor doesn't support optional zxdg_decoration_manager_v1!\n");
 
     if (!process_wayland.xdg_activation_v1)
-        ERR("Wayland compositor doesn't support xdg_activation_v1! (Flash Window will not be supported)\n");
+        ERR("Wayland compositor doesn't support xdg_activation_v1! (Window Activation will not be supported)\n");
+
+    if (!process_wayland.supports_win_pq)
+        ERR("Wayland compositor doesn't expose windows_bt2100 image description (HDR may look broken)!\n");
+
+    if (!process_wayland.supports_win_scrgb)
+        ERR("Wayland compositor doesn't expose windows_scrgb image description (HDR may look broken)!\n");
 
     process_wayland.initialized = TRUE;
 
diff --git a/dlls/winewayland.drv/wayland_output.c b/dlls/winewayland.drv/wayland_output.c
index 8a5e6b9a83f..070a96325ad 100644
--- a/dlls/winewayland.drv/wayland_output.c
+++ b/dlls/winewayland.drv/wayland_output.c
@@ -216,16 +216,8 @@ static void wayland_output_done(struct wayland_output *output)
         output->current.ref_lum = output->pending.ref_lum;
     }
 
-    output->current.supports_hdr = FALSE;
-
-    if (process_wayland.supports_extended_volume &&
-        process_wayland.supports_pq &&
-        process_wayland.supports_scrgb)
-    {
-        if (!process_wayland.supports_win_pq)
-            ERR("Compositor doesn't supports windows_bt2100, HDR10 ST2084 may look broken\n");
-        output->current.supports_hdr = (output->current.max_target_lum > output->current.ref_lum);
-    }
+    output->current.supports_hdr = process_wayland.supports_win_scrgb &&
+                                    (output->current.max_target_lum > output->current.ref_lum);
 
     output->pending_flags = 0;
 
@@ -801,65 +793,3 @@ void wayland_output_use_xdg_extension(struct wayland_output *output)
     zxdg_output_v1_add_listener(output->zxdg_output_v1, &zxdg_output_v1_listener,
                                 output);
 }
-
-static void wayland_color_manager_handle_supported_intent(void *data,
-    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t intent)
-{
-}
-
-static void wayland_color_manager_handle_supported_feature(void *data,
-    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t feature)
-{
-    pthread_mutex_lock(&process_wayland.output_mutex);
-
-    TRACE("feature %u\n", feature);
-
-    if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_SCRGB)
-        process_wayland.supports_scrgb = TRUE;
-    else if (feature == WP_COLOR_MANAGER_V1_FEATURE_WINDOWS_BT2100)
-        process_wayland.supports_win_pq = TRUE;
-    else if (feature == WP_COLOR_MANAGER_V1_FEATURE_EXTENDED_TARGET_VOLUME)
-        process_wayland.supports_extended_volume = TRUE;
-
-    pthread_mutex_unlock(&process_wayland.output_mutex);
-}
-
-static void wayland_color_manager_handle_supported_named_tf(void *data,
-    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t tf)
-{
-    pthread_mutex_lock(&process_wayland.output_mutex);
-
-    TRACE("named tf %u\n", tf);
-
-    if (tf == WP_COLOR_MANAGER_V1_TRANSFER_FUNCTION_ST2084_PQ)
-        process_wayland.supports_pq = TRUE;
-
-    pthread_mutex_unlock(&process_wayland.output_mutex);
-}
-
-static void wayland_color_manager_handle_supported_primaries(void *data,
-    struct wp_color_manager_v1 *wp_color_manager_v1, uint32_t primaries)
-{
-}
-
-static void wayland_color_manager_handle_done(void *data,
-                        struct wp_color_manager_v1 *wp_color_manager_v1)
-{
-}
-
-static const struct wp_color_manager_v1_listener wp_color_manager_listener = {
-    wayland_color_manager_handle_supported_intent,
-    wayland_color_manager_handle_supported_feature,
-    wayland_color_manager_handle_supported_named_tf,
-    wayland_color_manager_handle_supported_primaries,
-    wayland_color_manager_handle_done
-};
-
-/**********************************************************************
- *          wayland_color_manager_init
- */
-void wayland_color_manager_init(void)
-{
-    wp_color_manager_v1_add_listener(process_wayland.wp_color_manager_v1,
-                                     &wp_color_manager_listener, NULL);
-}
diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index 3ffa8150b98..257b91356ec 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1750,6 +1750,7 @@ void wayland_client_surface_attach_image_description(struct client_surface *clie
                                                      struct wp_image_description_v1 *image_desc)
 {
     struct wayland_client_surface *surface = impl_from_client_surface(client);
+
     if (!image_desc)
     {
         if (surface->wp_color_management_surface_v1)
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index a7c208bc33e..2df573ae285 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -268,10 +268,8 @@ struct wayland
     /* Protects the output_list, output_info_array, and the wayland_output.current states. */
     pthread_mutex_t output_mutex;
     LONG input_serial;
-    BOOL supports_pq;
-    BOOL supports_scrgb;
+    BOOL supports_win_scrgb;
     BOOL supports_win_pq;
-    BOOL supports_extended_volume;
 };
 
 struct wayland_output_mode
diff --git a/dlls/winex11.drv/vulkan.c b/dlls/winex11.drv/vulkan.c
index 12c42509245..ad5f0dd396c 100644
--- a/dlls/winex11.drv/vulkan.c
+++ b/dlls/winex11.drv/vulkan.c
@@ -70,9 +70,9 @@ static VkResult X11DRV_vulkan_surface_create( HWND hwnd, BOOL raw, const struct
     return VK_SUCCESS;
 }
 
-static VkResult X11DRV_vulkan_colorspace_configure(  VkColorSpaceKHR *colorspace, struct client_surface *client )
+static VkColorSpaceKHR X11DRV_vulkan_map_colorspace( VkColorSpaceKHR colorspace, struct client_surface *client )
 {
-    return VK_SUCCESS;
+    return colorspace;
 }
 
 static VkBool32 X11DRV_get_physical_device_presentation_support( struct vulkan_physical_device *physical_device, uint32_t index )
@@ -104,7 +104,7 @@ static void X11DRV_map_device_extensions( struct vulkan_device_extensions *exten
 static const struct vulkan_driver_funcs x11drv_vulkan_driver_funcs =
 {
     .p_vulkan_surface_create = X11DRV_vulkan_surface_create,
-    .p_vulkan_colorspace_configure = X11DRV_vulkan_colorspace_configure,
+    .p_vulkan_map_colorspace = X11DRV_vulkan_map_colorspace,
     .p_get_physical_device_presentation_support = X11DRV_get_physical_device_presentation_support,
     .p_map_instance_extensions = X11DRV_map_instance_extensions,
     .p_map_device_extensions = X11DRV_map_device_extensions,
diff --git a/include/wine/vulkan_driver.h b/include/wine/vulkan_driver.h
index 3bf3688880f..33ec7aef655 100644
--- a/include/wine/vulkan_driver.h
+++ b/include/wine/vulkan_driver.h
@@ -366,7 +366,7 @@ struct client_surface;
 struct vulkan_driver_funcs
 {
     VkResult (*p_vulkan_surface_create)(HWND, BOOL, const struct vulkan_instance *, VkSurfaceKHR *, struct client_surface **);
-    VkResult (*p_vulkan_colorspace_configure)( VkColorSpaceKHR *, struct client_surface * );
+    VkColorSpaceKHR (*p_vulkan_map_colorspace)( VkColorSpaceKHR, struct client_surface * );
     VkBool32 (*p_get_physical_device_presentation_support)(struct vulkan_physical_device *, uint32_t);
     void (*p_map_instance_extensions)( struct vulkan_instance_extensions *extensions );
     void (*p_map_device_extensions)( struct vulkan_device_extensions *extensions );
-- 
2.55.0

